csclient/csclient.go

```diff
@@ -249,31 +290,122 @@ func (c *Client) UploadResource(id *charm.URL, name, path string, file io.ReadSe
 	url := fmt.Sprintf("/%s/resource/%s?hash=%s&filename=%s", id.Path(), name, hash, path)
 
 	// Send the request.
-	resp, err := c.DoWithBody(req, url, file)
+	resp, err := c.DoWithBody(req, url, io.NewSectionReader(file, 0, size))
 	if err != nil {
-		return -1, errgo.NoteMask(err, "cannot post resource", isAPIError)
+		return 0, errgo.NoteMask(err, "cannot post resource", isAPIError)
 	}
 	defer resp.Body.Close()
 
 	// Parse the response.
 	var result params.ResourceUploadResponse
 	if err := parseResponseBody(resp.Body, &result); err != nil {
-		return -1, errgo.Mask(err)
+		return 0, errgo.Mask(err)
```

Comment:
Should we still call Finalizing() here, so that clients don't need to guess too much about whether this is multiplart vs singlepart?